solls/LanguageServer.cpp

```diff
@@ -0,0 +1,679 @@
+		{
+			m_currentNode = &_node;
+			return true;
+		}
+		return false;
+	}
+};
+
+// TOOD: maybe use SimpleASTVisitor here, if that would be a simple free-fuunction :)
+class ReferenceCollector: public frontend::ASTConstVisitor
+{
+private:
+	frontend::Declaration const& m_declaration;
+	std::vector<lsp::Server::DocumentHighlight> m_result;
+
+public:
+	explicit ReferenceCollector(frontend::Declaration const& _declaration):
+		m_declaration{_declaration}
+	{
+		fprintf(stderr, "finding refs: '%s', '%s'\n",
+				_declaration.name().c_str(),
+				_declaration.location().text().c_str());
+	}
+
+	std::vector<lsp::Server::DocumentHighlight> take() { return std::move(m_result); }
+
+	bool visit(Identifier const& _identifier) override
+	{
+		if (auto const declaration = _identifier.annotation().referencedDeclaration; declaration)
+			if (declaration == &m_declaration)
+				addReference(_identifier.location(), "(identifier)");
+
+		return visitNode(_identifier);
+	}
+
+	void addReference(SourceLocation const& _location, char const* msg = "")
+	{
+		auto const [startLine, startColumn] = _location.source->translatePositionToLineColumn(_location.start);
+		auto const [endLine, endColumn] = _location.source->translatePositionToLineColumn(_location.end);
+		auto const locationRange = lsp::Range{
+			lsp::Position{startLine, startColumn},
+			lsp::Position{endLine, endColumn}
+		};
+
+		fprintf(stderr, " -> found reference %s at %d:%d .. %d:%d\n",
+			msg,
+			startLine, startColumn,
+			endLine, endColumn
+		);
+
+		auto highlight = lsp::Server::DocumentHighlight{};
+		highlight.range = locationRange;
+		highlight.kind = lsp::Server::DocumentHighlightKind::Text; // TODO: are you being read or written to?
+
+		m_result.emplace_back(highlight);
+	}
+
+	// TODO: MemberAccess
+
+	bool visitNode(ASTNode const& _node) override
+	{
+		if (&_node == &m_declaration)
+		{
+			if (auto const* decl = dynamic_cast<Declaration const*>(&_node))
+				addReference(decl->nameLocation(), "(visitNode)");
+			else
+				addReference(_node.location(), "(visitNode)");
+		}
+
+		return true;
+	}
+};
+
+} // }}} end helpers
+
+LanguageServer::LanguageServer(lsp::Transport& _client, Logger _logger):
+	lsp::Server(_client, std::move(_logger)),
+	m_vfs()
+{
+}
+
+void LanguageServer::shutdown()
+{
+	logInfo("LanguageServer: shutdown requested");
+}
+
+LanguageServer::InitializeResponse LanguageServer::initialize(
+	string _rootUri,
+	map<string, string> _settings,
+	Trace _trace,
+	vector<WorkspaceFolder> _workspaceFolders
+)
+{
+	(void) _trace; // TODO: debuglog based on this config
+	(void) _settings; // TODO: user settings (such as EVM version)
+
+#if !defined(NDEBUG)
+	ostringstream msg;
+	msg << "LanguageServer: rootUri : " << _rootUri << endl;
+	for (auto const& workspace: _workspaceFolders)
+		msg << "                workspace folder: " << workspace.name << "; " << workspace.uri << endl;
+#endif
+
+	if (boost::starts_with(_rootUri, "file:///"))
```

Comment:
Why the triple-slash? Does this work on windows?
Also, do we add the `file://` prefix in the other file just to remove it here again?

Reply:
It does work on windows, because file URLs are always forward slashed. the tripple slash means I am expecting a path component that starts with a slash, i.e. a fully qualified path from disk root. ( I should test the windows case though , I've added that to my todo).